### DISCUSSION TITLE
Proper way to store Supabase client in Hono Cloudflare Workers

### DISCUSSION THREAD

**Discussion by bulutg** [honojs/hono#3718] [Q&A] [answered]
I'm building a Cloudflare Worker using Hono and Supabase, and I'm trying to understand the best way to handle the Supabase client initialization. Specifically, I want to know whether storing the Supabase client in Hono's context is appropriate, given how contexts are handled in the Cloudflare Workers environment.

Are Hono contexts generated for every request or are they maintained at the machine/worker level? This affects whether I should initialize the Supabase client for each request or keep a single instance.

I know I can store the Supabase client in the context like this:

```typescript 
import { Hono } from 'hono'
import { createClient } from '@supabase/supabase-js'

type Variables = {
    supabase: SupabaseClient
}

const app = new Hono<{ Variables: Variables }>()

app.use('*', async (c, next) => {
    // Is this the right approach?
    const supabase = createClient(
        'SUPABASE_URL',
        'SUPABASE_KEY'
    )
    c.set('supabase', supabase)
    await next()
})
```

What I'm unsure about:

- Is this creating a new Supabase client for every request?

- If so, is there a more efficient way to handle this in Cloudflare Workers?

- What's the recommended pattern for sharing database clients in Hono?

**Comment by marcomuser** [comment, marked as answer]
I find this little middleware helper to be very useful for this specific usecase: https://github.com/maou-shonen/hono-simple-DI. It allows you to define whether your dependency should be request scoped or reused across requests.

**Reply by bulutg** [reply]
Thanks! Could there be a race condition when multiple requests hit the service in "default scope", since they're sharing a single instance?

**Reply by marcomuser** [reply]
Not sure I yet understand what you mean. Are you worried about race conditions writing to the supabase database? Then yes, there could be race conditions. If you e.g. have usecases where several users write to the same resource simultaneously you need to handle conflict resolution in your application. Whether your database client is a singleton or not should not make a difference in this regard.
